### ISSUE TITLE
Compiler error using redefined custom types

### ISSUE THREAD

**Issue by wildchild04** [gleam-lang/gleam#3575]
## Description

I'm learning gleam by writing the good old `Thorsten Ball's "Writing An Interpreter In Go" `. at some point I wrote a custom type 
```gleam
type LiteralExpressionWrap {
  IntWrap(Int)
  StringWrap(String)
  BooleanWrap(Bool)
  StringWrap(String)
}
```

and this code crashed the LSP and  `gleam check` is returning the following message:
```
error: Fatal compiler bug!

This is a bug in the Gleam compiler, sorry!

Please report this crash to https://github.com/gleam-lang/gleam/issues/new
and include this error message with your report.

Panic: compiler-core/src/exhaustiveness.rs:746
        Custom type variants must exist: Type { name: "LiteralExpressionWrap", hint: AlternativeTypes(["Int", "Bool", "BitArray", "List", "String", "LiteralExpressionWrap", "UtfCodepoint", "Result", "Float", "Nil"]) }
Gleam version: 1.3.0
Operating system: linux

If you can also share your code and say what file you were editing or any
steps to reproduce the crash that would be a great help.

You may also want to try again with the `GLEAM_LOG=trace` environment
variable set.
```

I tried to isolate the issue by trying  to define a custom type in the same fashion just down below the other type definition

```gleam
type LiteralExpressionWrap {
  IntWrap(Int)
  StringWrap(String)
  BooleanWrap(Bool)
  // StringWrap(String)
}

type Bug {
  IntBug(Int)
  StringBug(String)
  BooleanBug(Bool)
  StringBug(String)
}
```

but this time I have the correct error and the expected behavior happening when I run `gleam check`

```
error: Duplicate definition
    ÔöîÔöÇ /home/wildchild/gleam/interpreter/test/parser/parser_test.gleam:284:3
    Ôöé
284 Ôöé   StringBug(String)
    Ôöé   ^^^^^^^^^^^^^^^^^ First defined here
285 Ôöé   BooleanBug(Bool)
286 Ôöé   StringBug(String)
    Ôöé   ^^^^^^^^^^^^^^^^^ Redefined here
```

I can reproduce this issue in Windows WLS running ubuntu and also in Mac OS using gleam 1.3
* Ubuntu 22.04.4 LTS on Windows 10 x86_64
* Sonoma 14.6.1

### Bug Reproduction Steps
1. Get the code from:
https://github.com/wildchild04/gleam-interpreter/tree/gleam-compiler-bug 
2. make sure you are in the branch `gleam-compiler-bug`
3. run `gleam check`

**Comment by giacomocavalieri** [comment]
Hello! Thanks for the bug report! It looks like this is a duplicate of https://github.com/gleam-lang/gleam/issues/3506

The fix will ship with Gleam v1.5 :)

**Comment by wildchild04** [comment]
awesome, then I'll close this one. Thanks!
